flink-yarn/src/test/java/org/apache/flink/yarn/ContainerRequestReflectorTest.java

```diff
@@ -0,0 +1,173 @@
+        Priority priority = Priority.newInstance(1);
+
+        AMRMClient.ContainerRequest containerRequest =
+                ContainerRequestReflector.INSTANCE.getContainerRequest(resource, priority, "GPU");
+        assertTrue(containerRequest instanceof AMRMClient.ContainerRequest);
+
+        containerRequest =
+                ContainerRequestReflector.INSTANCE.getContainerRequest(resource, priority, null);
+        assertTrue(containerRequest instanceof AMRMClient.ContainerRequest);
+
+        containerRequest =
+                ContainerRequestReflector.INSTANCE.getContainerRequest(resource, priority, "");
+        assertTrue(containerRequest instanceof AMRMClient.ContainerRequest);
+    }
+
+    @Test
+    public void testGetContainerRequestWithYarnSupport()
+            throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
+        assumeTrue(HadoopUtils.isMinHadoopVersion(2, 6));
+
+        Resource resource = Resource.newInstance(100, 1);
+        Priority priority = Priority.newInstance(1);
+
+        AMRMClient.ContainerRequest containerRequest =
+                ContainerRequestReflector.INSTANCE.getContainerRequest(resource, priority, "GPU");
+        assertTrue(containerRequest instanceof AMRMClient.ContainerRequest);
+        assertEquals("GPU", getNodeLabelExpressionWithReflector(containerRequest));
+
+        containerRequest =
+                ContainerRequestReflector.INSTANCE.getContainerRequest(resource, priority, null);
+        assertTrue(containerRequest instanceof AMRMClient.ContainerRequest);
+        assertNull(getNodeLabelExpressionWithReflector(containerRequest));
+
+        containerRequest =
+                ContainerRequestReflector.INSTANCE.getContainerRequest(resource, priority, "");
+        assertTrue(containerRequest instanceof AMRMClient.ContainerRequest);
+        assertNull(getNodeLabelExpressionWithReflector(containerRequest));
+    }
+
+    private String getNodeLabelExpressionWithReflector(AMRMClient.ContainerRequest containerRequest)
+            throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
+        Method method = containerRequest.getClass().getMethod("getNodeLabelExpression");
+        return (String) method.invoke(containerRequest);
+    }
+
+    /** Class which does not have required constructor. */
+    private static class ContainerRequestWithoutConstructor extends AMRMClient.ContainerRequest {
+
+        public ContainerRequestWithoutConstructor(
+                Resource capability, String[] nodes, String[] racks, Priority priority) {
+            super(capability, nodes, racks, priority);
+        }
+    }
+
+    /**
+     * Class which has constructor with the same signature as {@link
+     * org.apache.hadoop.yarn.client.api.AMRMClient.ContainerRequest} in Hadoop 2.6+.
+     */
+    private static class ContainerRequestWithConstructor extends AMRMClient.ContainerRequest {
+        private boolean relaxLocality;
```

Comment:
nit: Just remove this field at the moment. Once we need it, we can add this back.